codesigndocuitests/codesigndoc.go

```diff
@@ -0,0 +1,164 @@
+	if err != nil {
+		return false, false, fmt.Errorf("failed to list installed provisioning profiles, error: %s", err)
+	}
+
+	log.Debugf("Installed profiles:")
+	for _, profileInfo := range profiles {
+		log.Debugf(profileInfo.String(certificates...))
+	}
+
+	certificatesToExport, profilesToExport, err := getFilesToExport(buildPath, certificates, profiles, certificatesOnly)
+	if err != nil {
+		return false, false, err
+	}
+
+	// export code sign settings
+	fmt.Println()
+	fmt.Println()
+	log.Printf("🔦  Analyzing the builded artifacts, to get code signing settings...")
+
+	if err := codesign.CollectAndExportIdentities(certificatesToExport, outputDirPath, askForPassword); err != nil {
+		return false, false, err
+	}
+
+	if err := codesign.CollectAndExportProvisioningProfiles(profilesToExport, outputDirPath); err != nil {
+		return false, false, err
+	}
+
+	provProfilesUploaded := (len(profilesToExport) == 0)
+	certsUploaded := (len(certificatesToExport) == 0)
+
+	var shouldUpload bool
+	if !certificatesOnly {
+		fmt.Println()
+		shouldUpload, err = goinp.AskForBoolFromReader("Do you want to upload the provisioning profiles and certificates to Bitrise?", os.Stdin)
+		if err != nil {
+			return false, false, err
+		}
+	} else {
+		shouldUpload, err = goinp.AskForBoolFromReader("Do you want to upload the certificates to Bitrise?", os.Stdin)
+		if err != nil {
+			return false, false, err
+		}
+	}
+
+	if shouldUpload {
+		certsUploaded, provProfilesUploaded, err = bitriseio.UploadCodesigningFiles(certificatesToExport, profilesToExport, certificatesOnly, outputDirPath)
+		if err != nil {
+			return false, false, err
+		}
+	}
+
+	fmt.Println()
+	log.Successf("Exports finished you can find the exported files at: %s", outputDirPath)
+
+	if err := command.RunCommand("open", outputDirPath); err != nil {
+		log.Errorf("Failed to open the export directory in Finder: %s", outputDirPath)
+	} else {
+		fmt.Println("Opened the directory in Finder.")
+	}
+
+	return certsUploaded, provProfilesUploaded, nil
+}
+
+func getFilesToExport(buildPath string, installedCertificates []certificateutil.CertificateInfoModel, installedProfiles []profileutil.ProvisioningProfileInfoModel, certificatesOnly bool) ([]certificateutil.CertificateInfoModel, []profileutil.ProvisioningProfileInfoModel, error) {
+	certificatesToExport := []certificateutil.CertificateInfoModel{}
+	profilesToExport := []profileutil.ProvisioningProfileInfoModel{}
+
+	if certificatesOnly {
+		exportCertificate, err := collectExportCertificate(installedCertificates)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		certificatesToExport = append(certificatesToExport, exportCertificate...)
+	} else {
+		testRunners, err := NewIOSTestRunners(buildPath)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		for _, testRunner := range testRunners {
+			certsToExport, profsToExport, err := collectCertificatesAndProfiles(*testRunner, installedCertificates, installedProfiles, certificatesToExport, profilesToExport)
```

Comment:
what is the reason for this solution?
- collectCertificatesAndProfiles receives certificatesToExport
- collectCertificatesAndProfiles appends the test run's certificates to certificatesToExport and returns it
- later you add the returned list to certificatesToExport

Reply:
👍